File: BM5_合并k个已排序的链表/python_solutions/heap_solution.py
from queue import PriorityQueue

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def mergeKLists(self , lists: List[ListNode]) -> ListNode:
        min_heap = PriorityQueue()
        for idx in range(len(lists)):
            if lists[idx] != None:
                min_heap.put((lists[idx].val, idx))
                lists[idx] = lists[idx].next
    
        res = ListNode(-1)
        head = res
        
        while not min_heap.empty():
            val, idx = min_heap.get()

            head.next = ListNode(val)
            head = head.next

            if lists[idx] != None:
                min_heap.put((lists[idx].val, idx))
                lists[idx] = lists[idx].next
        
        return res.next

File: BM5_合并k个已排序的链表/java_solutions/HeapSolution.java
package BM5_合并k个已排序的链表.java_solutions;

import java.util.*;

public class HeapSolution {
    /**
     * 合并 k 个升序的链表并将结果作为一个升序的链表返回其头节点。
     *
     *
     * @param lists ListNode类ArrayList
     * @return ListNode类
     */
    public ListNode mergeKLists(ArrayList<ListNode> lists) {
        PriorityQueue<ListNode> pq = new PriorityQueue<ListNode>(
            (n1, n2) -> n1.val - n2.val
        );

        for (ListNode listNode : lists) {
            if(listNode != null) pq.add(listNode);
        }

        ListNode res = new ListNode(-1);
        ListNode head = res;

        while (pq.size() > 0) {
            ListNode min = pq.poll();
            
            head.next = min;
            head = min;

            if (min.next != null) pq.add(min.next);
        }

        return res.next;
    }
}